File: frontend/src/types/job.ts
export type JobStatus = "pending" | "running" | "completed" | "failed" | "cancelled";
export type JobStage = "ec2_starting" | "pipeline_running" | "snakemake_running" | "done" | null;
export type Tier = "small" | "medium" | "large";

export interface VcfVariant {
  chrom: string;
  pos: number;
  ref: string;
  alt: string;
  qual?: string;
  filter?: string;
}

export interface ResultFile {
  name: string;
  path: string;
  size_bytes?: number;
  url?: string;
  mime_type?: string;
  description?: string;
}

export interface Provenance {
  pipeline?: string;
  pipeline_version?: string;
  genome?: string;
  params?: Record<string, unknown>;
  n_samples?: number;
  completed_at: string;
  instance_type: string;
  runtime_seconds: number;
}

export interface JobResult {
  type?: "table" | "vcf" | "html_report" | "text" | "files";
  _mock?: boolean;
  provenance?: Provenance;
  // table
  columns?: string[];
  rows?: Record<string, string | number>[];
  // vcf
  variants?: VcfVariant[];
  // html_report
  html?: string;
  // text
  content?: string;
  // files
  files?: ResultFile[];
  // always present
  instance_type: string;
  runtime_seconds: number;
}

export interface JobListItem {
  job_id: string;
  status: JobStatus;
  stage: JobStage;
  tier: Tier;
  estimated_cost_usd: number;
  pipeline_id?: string | null;
  job_name?: string | null;
  created_at: string;
}

export interface Job {
  job_id: string;
  status: JobStatus;
  stage: JobStage;
  tier: Tier;
  estimated_cost_usd: number;
  pipeline_id?: string | null;
  job_name?: string | null;
  created_at?: string;
  result: JobResult | null;
  error: string | null;
}

export interface EstimateResponse {
  tier: Tier;
  instance_type: string;
  estimated_cost_usd: number;
  rationale: string;
  estimated_hours: number;
  pipeline_description: string;
}

export interface CheckoutResponse {
  checkout_url: string;
  session_id: string;
}

export interface CheckoutRequest {
  storage_key: string;
  file_type: string;
  tier: Tier;
  pipeline_id: string | null;
  estimated_cost_usd: number;
  n_samples: number;
  storage_key_r2?: string | null;
  workflow_config?: unknown | null;
}

export interface PresignResponse {
  upload_url: string;
  storage_key: string;
  recommended_tier: Tier;
  estimated_cost_usd: number;
  tier_rationale: string;
}

File: backend/app/schemas/job.py
from datetime import datetime
from typing import Any, Optional
from pydantic import BaseModel


class JobCreate(BaseModel):
    storage_key: str
    file_type: str              # fastq | bam (or any nf-core-accepted format)
    tier: str                   # small | medium | large
    estimated_cost_usd: float
    pipeline_id: Optional[str] = None   # sarek | snakemake | bioscript | custom-{tool}
    storage_key_r2: Optional[str] = None  # R2 read for paired-end FASTQ
    workflow_config: Optional[Any] = None  # Snakemake/BioScript workflow configuration
    job_name: Optional[str] = None        # Human-readable name set by user


# ── Result schemas ────────────────────────────────────────────────────────


class VcfVariant(BaseModel):
    chrom: str
    pos: Any
    id: str
    ref: str
    alt: str
    qual: Any
    filter: str
    info: str


class ResultFile(BaseModel):
    name: str
    path: str
    size_bytes: Optional[int] = None
    mime_type: Optional[str] = None
    description: Optional[str] = None


class JobResult(BaseModel):
    """Generalised result payload.

    The ``type`` field is used by the frontend ResultsPanel to select
    the appropriate renderer.  All payload fields are optional to allow
    partial results and future extension.
    """

    type: Optional[str] = None          # "table" | "vcf" | "html_report" | "text" | "files"

    # Generic table / count matrix (type="table")
    columns: Optional[list[str]] = None
    rows: Optional[list[dict[str, Any]]] = None

    # VCF variants (type="vcf")
    variants: Optional[list[VcfVariant]] = None

    # HTML report, e.g. MultiQC (type="html_report")
    html: Optional[str] = None

    # Plain text (type="text")
    content: Optional[str] = None

    # File list (type="files")
    files: Optional[list[ResultFile]] = None

    # Common metadata (always present)
    instance_type: str = ""
    runtime_seconds: int = 0


# ── Job response ──────────────────────────────────────────────────────────


class JobResponse(BaseModel):
    job_id: str
    status: str
    stage: Optional[str]
    tier: str
    estimated_cost_usd: float
    pipeline_id: Optional[str] = None
    storage_key_r2: Optional[str] = None
    workflow_config: Optional[Any] = None
    job_name: Optional[str] = None
    created_at: Optional[datetime] = None
    result: Optional[Any]
    error: Optional[str]

    model_config = {"from_attributes": True}


class JobListResponse(BaseModel):
    job_id: str
    status: str
    stage: Optional[str]
    tier: str
    estimated_cost_usd: float
    pipeline_id: Optional[str] = None
    job_name: Optional[str] = None
    created_at: datetime

    model_config = {"from_attributes": True}
